[RFC][WIP] Dmic nhlt parse

juimonen opened the pull request:
Idea here is to tap into topology2 pre-processing to generate nhlt blob containing endpoint descriptors, which include intel vendor specific blobs -> and add that to topology manifest. The nhlt blob generated should have the same structure as you would get from "cat /sys/firmware/acpi/tables/NHLT". Currently only dmic blob is implemented and I've been able to use that to configure dmic dai through kernel. SSP file is included but doesn't do anything yet.

lgirdwood reviewed (commented):
Really good work @juimonen - mostly needs a cleanup and more comments around the code blocks to describe what we are doing. It really has to be clear from comments too so everyone who has never heard of NHLT will know what its doing and also why.

juimonen commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+        uint16_t polarity_mic_b;
+        uint16_t clk_edge;
+        uint16_t skew;
+};
```
these are only used for storage for parameters from topology before actual blob calculation, so no memcpy's on these. I've packed the one's I'm building the final blob

juimonen commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+	uint32_t out_gain_left;
+	uint32_t out_gain_right;
+	uint32_t reserved2[2];
+} __attribute__((packed));
```
ok

juimonen commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+	uint32_t channel_ctrl_mask;
+	uint32_t chan_ctrl_cfg[2];
+	uint32_t channel_pdm_mask;
+	struct dmic_intel_pdm_ctrl_cfg pdm_ctrl_cfg[2];
```
ok need to use define

juimonen commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+#define MAX_NHLT_BUFFER_SIZE 25000
+char nhlt_string_buffer[MAX_NHLT_BUFFER_SIZE];
+
+static char * print_32_bit_as_bytes(char *dst, uint32_t val)
```
this is only for debug, but similar function in nhlt.c, and should be commented. It is basically the hex format alsa conf file needs for the "byte" field. So bytes printed as a string of hex values

juimonen commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.chan_ctrl_cfg[0]);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.chan_ctrl_cfg[1]);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.channel_pdm_mask);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].cic_control);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].cic_config);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].mic_control);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].pdmsm);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].reuse_fir_from_pdm);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[0].fir_control);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[0].fir_config);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[0].dc_offset_left);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[0].dc_offset_right);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[0].out_gain_left);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[0].out_gain_right);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[1].fir_control);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[1].fir_config);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[1].dc_offset_left);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[1].dc_offset_right);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[1].out_gain_left);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[0].fir_config[1].out_gain_right);
+	nhlt_p = print_32_bit_as_bytes(nhlt_p, dmic_vendor_blob.pdm_ctrl_cfg[1].pdmsm);
```
hmm sure... let me think if remove the whole debug

juimonen replied:
removed the whole thing

juimonen commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+	nhlt_p--;
+	*nhlt_p = '\0';
+
+	printf("%s\n", nhlt_string_buffer);
```
should be removed... old debug fuzz

juimonen commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+	return ret;
+}
+
+static int configure_registers(struct dmic_calc_configuration *cfg)
```
it is not the spec that requires this.  I think it is totally voluntary, it is unspecified vendor blob. Closed fw wants the parameters in this format.

juimonen commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+	const char *id;
+	long int_val = 0;
+	int driver_version = 1;
+	int io_clk = 19200000;
+	int dai_index_t = 0;
+	int num_pdm_active = 1;
+	int fifo_word_length = 32;
+	int clk_min = 2400000;
+	int clk_max = 4800000;
+	int duty_min = 40;
+	int duty_max = 60;
+	int sample_rate = 48000;
+	int unmute_ramp_time_ms = 400;
```
oh yes, I need to polish the defaults processing

juimonen replied:
init function added for setting default values

juimonen commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+
+	/* top level struct */
+	memcpy(vendor_blob, &dmic_vendor_blob, 40);
+	vendor_blob += 40;
```
yes and I should. little bit struggling again with the variable size nested arrays...

juimonen commented on topology/intel/dmic-nhlt.h:
```diff
@@ -0,0 +1,32 @@
+			 snd_config_t *parent);
+
+int get_dmic_vendor_blob_size();
+void copy_dmic_vendor_blob(uint8_t *vendor_blob);
```
I was actually removing that, because strictly this is not nhlt spec, it is vendor blob. I wanted exactly to avoid confusion of the concepts. OTH I don't mind, so whatever is most clear thing to do

juimonen commented on topology/intel/dmic/dmic_macros.h:
```diff
@@ -0,0 +1,255 @@
+
+#define BIT(b)                  (1UL << (b))
+
+#define MASK(b_hi, b_lo)        \
+        (((1ULL << ((b_hi) - (b_lo) + 1ULL)) - 1ULL) << (b_lo))
+#define SET_BIT(b, x)           (((x) & 1) << (b))
+#define SET_BITS(b_hi, b_lo, x) \
+        (((x) & ((1ULL << ((b_hi) - (b_lo) + 1ULL)) - 1ULL)) << (b_lo))
+
+#define EXP_FIXED_INPUT_QY 27
+#define EXP_FIXED_OUTPUT_QY 20
+#define DB2LIN_FIXED_INPUT_QY 24
+#define DB2LIN_FIXED_OUTPUT_QY 20
```
hmm ok, I need to go through these and see what I can find in userspace -> if not, I'll prefix

juimonen commented on topology/intel/dmic/numbers.c:
```diff
@@ -0,0 +1,92 @@
+		val >>= 1;
+
+	return 31 - c;
+}
```
I should get warnings from not used functions and variables -> but will check

juimonen replied:
I just could not find very easy/clear user space substitutes for these

juimonen commented on topology/intel/dmic/numbers.h:
```diff
@@ -0,0 +1,96 @@
+
+#include <stdint.h>
+
+#define MIN(a, b) ({		\
+	typeof(a) __a = (a);	\
+	typeof(b) __b = (b);	\
+	__a > __b ? __b : __a;	\
+})
+
+#define MAX(a, b) ({		\
+	typeof(a) __a = (a);	\
+	typeof(b) __b = (b);	\
+	__a < __b ? __b : __a;	\
+})
+#define ABS(a) ({		\
+	typeof(a) __a = (a);	\
+	__a < 0 ? -__a : __a;	\
+})
+#define SGN(a) ({		\
+	typeof(a) __a = (a);	\
+	__a < 0 ? -1 :		\
+	__a > 0 ? 1 : 0;	\
+})
+
```
ok

juimonen commented on topology/intel/dmic/numbers.h:
```diff
@@ -0,0 +1,96 @@
+		c++;
+
+	return c;
+}
```
ok

juimonen commented on topology/intel/nhlt.h:
```diff
@@ -0,0 +1,31 @@
+
+#include <stdint.h>
+
+int set_dai_nhlt(struct tplg_pre_processor *tplg_pp,
```
ok these are the "public" functions visible upstairs... I can prefix these yes

juimonen commented on topology/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,860 @@
+#include "pre-processor.h"
+#include "ssp-nhlt.h"
+
+#define SOF_DAI_FMT_I2S         1 /**< I2S mode */
+#define SOF_DAI_FMT_RIGHT_J     2 /**< Right Justified mode */
+#define SOF_DAI_FMT_LEFT_J      3 /**< Left Justified mode */
+#define SOF_DAI_FMT_DSP_A       4 /**< L data MSB after FRM LRC */
+#define SOF_DAI_FMT_DSP_B       5 /**< L data MSB during FRM LRC */
+#define SOF_DAI_FMT_PDM         6 /**< Pulse density modulation */
+
+#define SOF_DAI_FMT_CONT        (1 << 4) /**< continuous clock */
+#define SOF_DAI_FMT_GATED       (0 << 4) /**< clock is gated */
+
+#define SOF_DAI_FMT_NB_NF       (0 << 8) /**< normal bit clock + frame */
+#define SOF_DAI_FMT_NB_IF       (2 << 8) /**< normal BCLK + inv FRM */
+#define SOF_DAI_FMT_IB_NF       (3 << 8) /**< invert BCLK + nor FRM */
+#define SOF_DAI_FMT_IB_IF       (4 << 8) /**< invert BCLK + FRM */
+
+#define SOF_DAI_FMT_CBP_CFP     (0 << 12) /**< codec bclk provider & frame provider */
+#define SOF_DAI_FMT_CBC_CFP     (2 << 12) /**< codec bclk consumer & frame provider */
+#define SOF_DAI_FMT_CBP_CFC     (3 << 12) /**< codec bclk provider & frame consumer */
+#define SOF_DAI_FMT_CBC_CFC     (4 << 12) /**< codec bclk consumer & frame consumer */
+
+#define SOF_DAI_FMT_FORMAT_MASK         0x000f
+#define SOF_DAI_FMT_CLOCK_MASK          0x00f0
+#define SOF_DAI_FMT_INV_MASK            0x0f00
+#define SOF_DAI_FMT_CLOCK_PROVIDER_MASK 0xf000
```
it is not uapi, so don't know what would be the best way here... Yes feels idiotic to redefine everything

lgirdwood replied:
Lets duplicate how ALSA lib does this with teh kernel ABI. IIRC, a header file is duplicated.

plbossart replied:
you want to remove references to SOF though, completely unrelated.

juimonen replied:
ok

juimonen commented on topology/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,860 @@
+
+	sscr0 = ssp_read(dai, SSCR0);
+
+#if CONFIG_INTEL_MN
```
sorry ssp still not even compiled, file just added, tried to explain that in Pr message

plbossart commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+#include "dmic-nhlt.h"
+#include "dmic/pdm_decim_table.h"
+#include "dmic/dmic_macros.h"
+
```
and add a pointer to the NHLT public spec.

juimonen replied:
pointer to spec added

plbossart commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+        uint32_t wake_up_time;
+        uint32_t min_clock_on_time;
+        uint32_t unmute_ramp_time;
+        struct dmic_config_pdm pdm[SOF_DAI_INTEL_DMIC_NUM_CTRL];
```
WHY SOF? This is not dependent on any SOF definitions or choices, and not even on SOF use.

juimonen replied:
removed

plbossart commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+struct dmic_calc_decim_modes {
+        int16_t clkdiv[DMIC_MAX_MODES];
+        int16_t mcic[DMIC_MAX_MODES];
+        int16_t mfir[DMIC_MAX_MODES];
```
INTEL_DMIC?

plbossart commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+	uint32_t pdmsm;
+	uint32_t reuse_fir_from_pdm;
+	uint32_t reserved1[2];
+	struct dmic_intel_fir_config fir_config[2];
```
2 -> define?

plbossart commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+	uint32_t reuse_fir_from_pdm;
+	uint32_t reserved1[2];
+	struct dmic_intel_fir_config fir_config[2];
+	/* uint32_t fir_coeffs[0]; */
```
remove?

juimonen replied:
ok

plbossart commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+
+struct dmic_intel_config_data {
+	uint32_t gateway_attributes;
+	uint32_t ts_group[4];
```
4 -> define?

plbossart commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+	dmic_vendor_blob.ts_group[1] = 4294967040;
+	dmic_vendor_blob.ts_group[2] = 4294967295;
+	dmic_vendor_blob.ts_group[3] = 4294967295;
+	dmic_vendor_blob.ts_group[3] = 4294967295;
```
what is all this?

juimonen replied:
removed... and added a better explanation

plbossart commented on topology/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1632 @@
+	const char *id;
+	long int_val = 0;
+	int driver_version = 1;
+	int io_clk = 19200000;
```
is this generic? This wouldn't work with 'older' platforms with a 24MHz clk.

juimonen replied:
removed default values.

plbossart commented on topology/intel/nhlt.c:
```diff
@@ -0,0 +1,474 @@
+ * https://uefi.org/sites/default/files/resources/ACPI_6_3_final_Jan30.pdf
+ *
+ * Idea is to generate similar blob as you would get from:
+ * 'cat /sys/firmware/acpi/tables/NHLT'
```
does this support the case with 2 mics on PDM1 (not PDM0)? we've had this configuration on some Lenovo platforms.

juimonen replied:
yes it should, in topology2 you need to disable pdm0 "mic_enable"s and enable pdm1 "mic_enable"

plbossart commented on topology/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,860 @@
+#define SSCR2_LJDFD		BIT(17)
+#define SSCR2_MMRATF		BIT(18)
+#define SSCR2_SMTATF		BIT(19)
+#endif
```
NHLT only supports CAVS, suggest removing all irrelevant references to older platforms, it's dead code

plbossart commented on topology/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,860 @@
+
+#if CONFIG_CAVS
+
+#include <sof/lib/clk.h>
```
??

plbossart commented on topology/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,860 @@
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
+  The full GNU General Public License is included in this distribution
+  in the file called LICENSE.GPL.
```
this file is taken from the SOF tree, why not keep the BSD-3 clause license? I don't see the point of changing to GPL, it's just confusing, isn't it?

plbossart commented on topology/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,860 @@
+	dmic_build_nhlt();
+
+	return 0;
+}
```
I think you should really split the code in two
a) new NHLT parts with the license of your choice (no objection to LGPL on my side)
b) helpers ripped from the SOF tree without a license change. that way you can add fixes later (both ways) on without having to worry about license, etc.

plbossart commented on topology/pre-process-object.c:
```diff
@@ -1054,7 +1072,8 @@ static const struct build_function_map *tplg_object_get_map(struct tplg_pre_proc
 		if (!strcmp(class_type, "Dai") &&
-		    !strcmp(object_build_map[i].class_type, "Dai"))
+		    !strcmp(object_build_map[i].class_type, "Dai")  &&
+			!strcmp(object_build_map[i].class_name, ""))
```
is this commit separated on purpose or could it be squashed with the previous? I don't see what this brings as a separate patch, or I don't have the background to understand it.

juimonen replied:
yeah, it is supposed to be separated. This is the part where I "tap" into @ranj063 's topology2 pre-processing. So without this none of the dmic/ssp bananas is used. So I wanted it to be separate commit so we could for example set in under some flag -> ok, do generate the nhlt blob or don't do it.

juimonen replied:
I now squashed the commits

plbossart reviewed (commented):
Random comments based on my superficial understanding of all things NHLT.

juimonen commented:
@plbossart thanks for the comments, they are good. Kind of nasty crossfire here as I need to use nhlt (kernel internal header) + sof dmic headers and calculations. Exactly I was thinking to replicate the code directly from sof, but it brings in files with many not needed macros and functions i.e. the dmic code for this part in sof is not done to be easily exported. And Liam wanted me to clean it up -> so it maybe comes cleaner/neater, but I will miss the portability with sof.  But I need to think about it, the separate licensing and exact replica of sof code kind of sounds good to me, because I need to fix/maintain this. 

The SSP is still in development, not even compiling it yet, tried to explain that poorly in PR comment.

juimonen commented:
updated with changes:
- refactored the file structure (nhlt spec processing in "intel" directory, dmic and ssp vendor specific under "intel/dmic" and "intel/ssp" 
- nhlt vendor blob parameters are now kept in pre-processor as private_data (no static data variables anymore)
- added function to initialize vendor blob parameters (when parsing starts -> init is called, during parsing -> internal structs are filled, when parsing ends -> nhlt acpi blob with endpoints is generated and added to manifest)
- dmic variable size blob support added (also for debug printing)
- ts_group and pdm_ctrl masks are taken into use
- added a lot of documentation/comments into the code
- ssp still not tested at all
- dmic needs hw ver3 support at some point (this blob is tested in tgl)

juimonen commented:
and I tried to close the comments I've addressed, those not addressed should be still open.

lgirdwood commented:
@beatabaranowska ping ?
@juimonen how is the SSP part coming along. When do you think it can be reviewed

juimonen commented:
@lgirdwood  basic SSP support is now in this updated PR. However, it is supporting the single duplex SSP endpoint we currently have in our test topologies (at least I 've been able to run SSP playback with it). I need to add support for multiple SSP endpoints and clean up the SSP code (mostly as I have some sof definitions there). So dmic has always 1 nhlt endpoint, but SSP can have several, hence the difference.

beatabaranowska commented:
@juimonen this is because in nhlt we provide dmic (the intel one) so one edpoint. and ssp depends the number of ports. currently 6 I guess.

beatabaranowska reviewed (commented):
just quick review. not finished yet.

juimonen commented on topology/intel/dmic/dmic-nhlt.h:
```diff
@@ -0,0 +1,32 @@
+int dmic_nhlt_set_params(struct tplg_pre_processor *tplg_pp, snd_config_t *cfg,
+			 snd_config_t *parent);
+int dmic_nhlt_get_vendor_blob_size(struct tplg_pre_processor *tplg_pp);
+void dmic_nhlt_get_vendor_blob(struct tplg_pre_processor *tplg_pp, uint8_t *vendor_blob);
```
hmm ok I'm fine changing the name to what is most "clear" here. In this case I associate the "vendor", "Intel" and "hw" to be kind of the same thing :) So this is a blob specific to intel dmic hw and the idea is just to differentiate the "secret" specific config (not defined in nhlt spec) from other structs.

juimonen commented on topology/intel/dmic/dmic_macros.h:
```diff
@@ -0,0 +1,281 @@
+#ifndef __DMIC_MACROS_H
+#define __DMIC_MACROS_H
+
+#define MIN(a, b) ({		\
```
well here it is macro used only by the dmic code :) Anyway these are macros that could be used in generic way. I have similar thing in SSP directory -> I can try to unify those. It could be that at some point here's a new endpoint that can use these macros. It is just about the code partitioning and where to "hide" the files so the directory would look clean.

juimonen commented on topology/intel/dmic/pdm_decim_int32_02_4323_5100_010_095.h:
```diff
@@ -0,0 +1,124 @@
+
+};
+
+struct pdm_decim pdm_decim_int32_02_4323_5100_010_095 = {
```
@singalsuo can you chime in here, as I will get it wrong anyway...

singalsu replied:
The header files contain a number of raw pre-computed low-pass filter coefficients set. They are not strictly hardware specific but assume the decimation HW filter topology to be similar (5th order CIC and direct form FIR). They include the CIC response compensation that is specific for the decimation factor. I call them raw because the computation in this PR scales them for the needed decimation gain. These are used by default and topology build does not need any specific math software.

If the topology specifies custom filter parameters (passband characterics, stopband characteristics) then the topology build will launch Octave to compute the filter for the given specification. This has not been yet implemented but it is straightforward. Need to have packages octave and octave-signal installed. Or alternatively Matlab with signal processing package.

juimonen commented on topology/intel/nhlt-internal.h:
```diff
@@ -0,0 +1,155 @@
+        uint16_t skew;
+};
+
+struct dmic_config_dai {
```
the generation in the whole thing happens in 2 parts: 
1) I gather the parameters from topology to intermediate structs (as they are parsed one-by-one)
2) Then the specific config blob for nhlt is calculated

So there are 2 types of structs in this file per endpoint: gathering the parameters and end result.

We could argue that none of these structs is defined in nhlt. I was trying to implicate with the file name that these are structs for "internal" calculation.

juimonen commented on topology/intel/nhlt.h:
```diff
@@ -0,0 +1,166 @@
+
+#define NHLT_VENDOR_ID_INTEL 0x8086
+
+#define NHLT_DEVICE_ID_INTEL_PDM_DMIC 0xAE20
```
sorry didn't get this... If I am creating the nhlt blob for Intel hw and dmic, shouldn't set these?

juimonen commented on topology/intel/nhlt.h:
```diff
@@ -0,0 +1,166 @@
+#define NHLT_DEVICE_TYPE_SSP_MODEM 2
+#define NHLT_DEVICE_TYPE_SSP_ANALOG 4
+#define NHLT_DEVICE_TYPE_PDM_CAVS18 0
+#define NHLT_DEVICE_TYPE_PDM_CAVS15 1
```
hmm yeah if nobody is using these, again there is DeviceType in the nhlt spec as subcategory for LinkType. But yes was also little bit confused what to set there...

singalsu commented on topology/intel/dmic/dmic-nhlt.c:
```diff
@@ -0,0 +1,1827 @@
+	tplg_pp_debug("printing dmic nhlt internal data");
+
+	/* top level struct */
+	tplg_pp_debug("gateway attributes: %u", nhlt->dmic_blob.gateway_attributes);
```
Nit, "0x%08x" would make a nicer format for debug output. Many of the bit fields are aligned at hex digit boundaries so hex is easier to decode by eye.

juimonen replied:
ok

singalsu commented on topology/intel/dmic/dmic-nhlt.c:
```diff
@@ -0,0 +1,1827 @@
+		remind = nhlt->dmic_fir_array.fir_len[0] % 8;
+		for (i = 0; i < lines; i++) {
+			line = i * 8;
+			tplg_pp_debug("%d %u %u %u %u %u %u %u %u", i,
```
These are signed integers so %d or hex would be better.

juimonen replied:
ok

singalsu commented on topology/intel/dmic/dmic-nhlt.c:
```diff
@@ -0,0 +1,1827 @@
+	 * topology if it is non-zero, otherwise use default length.
+	 */
+	/* unmute_ramp related stuff is not used in nhlt at all */
+	if (nhlt->dmic_prm[di].unmute_ramp_time)
```
I think you can remove lines 1187 - 1198.

juimonen replied:
ok

singalsu reviewed (changes requested):
Looks good, thanks! Some unnecessary code to remove, maybe improve debug prints.

singalsu commented on topology/intel/dmic/dmic-macros.h:
```diff
@@ -0,0 +1,262 @@
+/* Hardwired log ramp parameters. The first value is the initial gain in
+ * decibels. The second value is the default ramp time.
+ */
+#define LOGRAMP_START_DB Q_CONVERT_FLOAT(-90, DB2LIN_FIXED_INPUT_QY)
```
Sorry I didn't notice in earlier round, the LOGRAMP #defines are unnecessary.

juimonen replied:
ok

singalsu reviewed (approved):
Looks OK but please remove the rest of gain ramp related stuff.

lgirdwood commented on configure.ac:
```diff
@@ -3,7 +3,7 @@ AC_PREREQ(2.59)
 AC_CONFIG_SRCDIR([aplay/aplay.c])
 AC_PREFIX_DEFAULT(/usr)
-AM_INIT_AUTOMAKE
+AM_INIT_AUTOMAKE([subdir-objects])
```
Can we say why we need this in teh commit message.

lgirdwood replied:
I dont think you need to do this here, configure.ac should be able to individually list the subdirs (there is another macro for it)

lgirdwood commented on topology/Makefile.am:
```diff
@@ -9,9 +9,23 @@ endif
-		    pre-process-dapm.c pre-process-dai.c
-
-noinst_HEADERS = topology.h pre-processor.h
+		pre-process-dapm.c pre-process-dai.c intel/nhlt.c intel/dmic/dmic-nhlt.c \
```
Can commit message also contain a link to the NHLT spec. 
https://01.org/sites/default/files/595976_intel_sst_nhlt.pdf
We also need to state the use case, e.g. NHLT data will be build able via topology for firmware consumption.

lgirdwood commented on topology/intel/dmic/dmic-macros.h:
```diff
@@ -0,0 +1,241 @@
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
+  The full GNU General Public License is included in this distribution
+  in the file called LICENSE.GPL.
```
Can we make this dual licence with BSC 3c since its derived from BSD.

lgirdwood commented on topology/pre-processor.c:
```diff
@@ -264,6 +264,8 @@ int pre_process(struct tplg_pre_processor *tplg_pp, char *config, size_t config_
 	}
 
+	tplg_pp_config_debug(tplg_pp, tplg_pp->output_cfg);
+
```
to be squashed ?

lgirdwood reviewed (changes requested):
Can you split into more patches i.e. one for nhtl core, one for ssp, one for dmic. This will make it easier for Jaroslav to review.

We also need to explain very early on in the commit message that the NHLT blobs are really register programming values that will be consumed by firmware based on topology data. This explains why the code looks like a driver.

lgirdwood commented on topology/intel/nhlt-processor.c:
```diff
@@ -0,0 +1,291 @@
+#include "nhlt-internal.h"
+#include "nhlt.h"
+
+#define MAX_EPS_COUNT 20
```
EPS ?

lgirdwood reviewed (commented):
One thing that Jaroslav likes to see is a config example before each parse function as a comment i.e.
```c
/*
* This function parses the following config items.
*
* config_blah1 {
*     data1 [
*         item1
*     ]
* }
*/
int parse_blah1()
{}
``` 
Every function that parses any data should have this. If the parsing is done by other code than instead I would paste an example of each NHLT data item in topology format so Jaroslav can see whet is the input and how we convert to output.

Commit:
topology: add nhlt generation into topology2.0 pre-processing

Add optional generation of nhlt acpi table into topology2.0 processing.
Nhlt internal structure is defined in:
https://01.org/sites/default/files/595976_intel_sst_nhlt.pdf
Nhlt acpi table contain vendor specific binary data blobs that are used
in some Intel dsp platforms for configuring the dmic and ssp hardware.

The beef in this code is to generate the vendor specific binary blobs,
but as there is existing nhlt parser in kernel there's no point of
re-inventing the container: just use the existing nhlt acpi table
format. Basically this code is creating similar nhlt acpi table that you
would get from: cat /sys/firmware/acpi/tables/NHLT

Nhlt generation is done by adding auto attribute function for
topology2.0 dai processing, saving parameters and generating the nhlt
struct after parsing. Nhlt struct is added into the manifest as bytes
field.

Nhlt processing uses a private data in tplg_pre_processor struct and
needs to be initialized before parsing starts. After all dais have been
parsed the nhlt can be created. Nhtl structs are almost direct copy from
related kernel header. Vendor specific blob generation will be added in
later commits.

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

Commit:
topology: add Intel dmic vendor blob generation for nhlt

In some Intel platforms the nhlt vendor specific blobs are sent to the
dsp over ipc. Add the generation of Intel dmic blob into the nhlt code.

Dmic code is lifted from Sound Open Firmware (sof) code base, thus it
will have BSD-3 license.

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

Commit:
topology: add Intel ssp vendor blob generation for nhlt

In some Intel platforms the nhlt vendor specific blobs are sent to the
dsp over ipc. Add the generation of Intel ssp blob into the nhlt code.

Ssp code is lifted from Sound Open Firmware (sof) code base, thus it
will have BSD-3 license.

Signed-off-by: Jaska Uimonen <jaska.uimonen@linux.intel.com>

lgirdwood commented:
ping

juimonen commented:
new version:
- removed subdirs clause from configure
- separated into 3 patches -> 1) nhlt (lgpl), 2) dmic (bsd-3) 3) ssp (bsd-3)
- tried to put more examples into the comments, how things are parsed from topo2
- now includes vendor specific microphone geometry parsing (in nhlt spec)

still to check:
- consistent function naming
- magic array defines
- ssp clock source prcoessing

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,287 @@
+	return 0;
+}
+
+/* called at the beginning of topology per-processing */
```
typo pre-processing

juimonen replied:
ok

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,287 @@
+	if (!tplg_pp)
+		return -EINVAL;
+
+	/* set default values here */
```
is this aTODO?

juimonen replied:
no just to say at this point we initialize and set possible default values for internal structs... but can be removed

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,287 @@
+	return 0;
+}
+
+/* called at the end of topology per-processing */
```
same typo

juimonen replied:
ok

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,287 @@
+	int i;
+
+	nhlt = (struct intel_nhlt_params *)tplg_pp->private_data;
+
```
remove newline

juimonen replied:
ok

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,287 @@
+	tplg_pp_debug("create_nhlt_blob ep count %d total size %u", eps_count, nhlt_size);
+
+	/* allocate the bytes string, every byte is alsa conf byte string format "0xA1," */
+	bytes_string_buffer = calloc(nhlt_size * 5, sizeof(uint8_t));
```
sorry having a hard time understanding the *5, can you please explain why?

juimonen replied:
as the "bytes" needs to be a string of hex bytes separated by comma i.e "0xa1," -> 5 chars

plbossart replied:
add the answer above as a comment

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,287 @@
+	/* remove all enpoints */
+	for (i = 0; i < eps_count; i++) {
+		free(eps[i]);
+		eps[i] = NULL;
```
why is this needed?

juimonen replied:
you mean the NULL? can be removed.... free is there just for symmetry, process dies after compiling anyway

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,287 @@
+		free(eps[i]);
+		eps[i] = NULL;
+	}
+	eps_count = 0;
```
eps_count is a local variable. why do we need to reset this?

juimonen replied:
no need, will remove

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,287 @@
+	/* print the whole thing in hex characters */
+	print_nhlt_as_hex_bytes(&blob, eps, bytes_string_buffer);
+
+	add_to_manifest(tplg_pp, bytes_string_buffer);
```
missing ret = ? then you dont need to init ret

juimonen replied:
ok

ranj063 commented on topology/pre-process-object.c:
```diff
@@ -934,6 +935,23 @@ static int tplg_build_generic_object(struct tplg_pre_processor *tplg_pp, snd_con
+	if (ret < 0)
+		return ret;
+
+	tplg_pp_debug("entering dai auto attr id %s", id);
```
can we clean up the debug prints a bit to keep only the most critical ones. I understand these are useful for you while you're inplementing the feature but they may pollute the output when building with debug enabled for the end user

juimonen replied:
yes, will clean

ranj063 commented on topology/pre-process-object.c:
```diff
@@ -1051,7 +1069,8 @@ static const struct build_function_map *tplg_object_get_map(struct tplg_pre_proc
 		if (!strcmp(class_type, "Dai") &&
-		    !strcmp(object_build_map[i].class_type, "Dai"))
+		    !strcmp(object_build_map[i].class_type, "Dai")  &&
+			!strcmp(object_build_map[i].class_name, ""))
```
is this change really required?

juimonen replied:
actually not.... not sure why I added it originally. Will remove

ranj063 commented on topology/pre-processor.c:
```diff
@@ -162,6 +164,10 @@ static int pre_process_config(struct tplg_pre_processor *tplg_pp, snd_config_t *
 	}
 
+	/* create nhlt blob and add it to manifest */
+	if (nhlt_process)
```
I'm not sure it is a good idea to have the flag set during compilation. I think this should come from the conf file itself. For example, when processing the DMIC object, we should somehow be able to say add NHLT blob for this DMIC

juimonen replied:
hmm ok... yes could be done. let me try it out

ranj063 commented on topology/topology.c:
```diff
@@ -417,6 +420,13 @@ int main(int argc, char *argv[])
 		case 'p':
 			pre_process_config = true;
 			break;
+		case 'l':
```
right this is the option that should come from the conf file itself

juimonen replied:
now in manifest...

ranj063 commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+			      dmic->dmic_prm[0].pdm[i].enable_mic_a);
+
+		if (swap_check && swap[i])
+			ret = -1;
```
and break?

juimonen replied:
yes

ranj063 commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+		cnt >>= 1;
+		stereo[i] = cnt;
+
+		/* Swap channels if only mic B is used for mono processing. */
```
it would make it a lot easier to understand, if you had a check like:
```
if (stereo[i]) continue;
```
then the check for !cnt below becomes redundant

juimonen replied:
talking to Seppo, original had many revisions about this, with several contributors -> so conclusion maybe not touch this this time.

ranj063 commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+		}
+	}
+
+	tplg_pp_debug("");
```
remove?

juimonen replied:
yeah part of the debug saga, so probably most of these will go away

ranj063 commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+
+	if (di >= DMIC_HW_FIFOS) {
+		tplg_pp_debug("dmic_set_config(): dai->index exceeds number of FIFOs");
+		ret = -1;
```
same comment with the -1 here as well

juimonen replied:
ok

ranj063 commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+	if (!nhlt) {
+		nhlt = calloc(1, sizeof(struct intel_nhlt_params));
+		if (!nhlt)
+			return -EINVAL;
```
ENOMEM

juimonen replied:
yes

ranj063 commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+
+	dmic = calloc(1, sizeof(struct intel_dmic_params));
+	if (!dmic)
+		return -EINVAL;
```
ENOMEM

juimonen replied:
yes

ranj063 commented on topology/nhlt/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,696 @@
+	/* get default values from class definition */
+	ret = snd_config_search(tplg_pp->input_cfg, "Class.Dai.SSP", &class_cfg);
+	if (ret < 0)
+		return -EINVAL;
```
return ret?

juimonen replied:
ok

ranj063 commented on topology/nhlt/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,696 @@
+			if (snd_config_get_string(n, &tdm_padding_per_slot))
+				return -EINVAL;
+		}
+	}
```
we can avoid all these checks above if we provide default values for the SSP DAI class or make them mandatory right?

ranj063 commented on topology/nhlt/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,696 @@
+	tplg_pp_debug("default tdm_padding_per_slot: %s", tdm_padding_per_slot);
+	tplg_pp_debug("");
+
+	/* set instance soecific values */
```
specific

ranj063 replied:
the object values always take precedence over the class default values. so this shouldnt be needed if the object has already been processed

juimonen replied:
hmm ok, so has it been processed at this stage? so you mean possible default values are already set as object values?

ranj063 commented on topology/nhlt/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,696 @@
+	tplg_pp_debug("default hw_config rx_slots: %ld", rx_slots);
+	tplg_pp_debug("");
+
+	/* set object values */
```
same comment as above

ranj063 commented on topology/nhlt/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,696 @@
+	 */
+
+	ret = ssp_get_params(tplg_pp, dai_index, &virtualbus_id, &formats_count);
+
```
remove newline

juimonen replied:
ok

juimonen commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -215,6 +217,15 @@ int nhlt_create(struct tplg_pre_processor *tplg_pp)
 	/* add endpoints here, increase eps_count */
 
+	/* we always have only 0 or 1 dmic ep */
+	if (nhlt_dmic_get_ep_count(tplg_pp)) {
```
topology/nhlt/intel/dmic-nhlt.h

juimonen commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -215,6 +217,15 @@ int nhlt_create(struct tplg_pre_processor *tplg_pp)
+	/* we always have only 0 or 1 dmic ep */
+	if (nhlt_dmic_get_ep_count(tplg_pp)) {
+		/* the index is always 0 in dmic case */
+		ret = nhlt_dmic_get_ep(tplg_pp, &eps[eps_count], 0);
```
topology/nhlt/intel/dmic-nhlt.h

juimonen commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -215,6 +217,15 @@ int nhlt_create(struct tplg_pre_processor *tplg_pp)
+	/* we always have only 0 or 1 dmic ep */
+	if (nhlt_dmic_get_ep_count(tplg_pp)) {
+		/* the index is always 0 in dmic case */
+		ret = nhlt_dmic_get_ep(tplg_pp, &eps[eps_count], 0);
```
was just for symmetry because of dmic, can be removed

juimonen commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -282,6 +294,12 @@ int nhlt_set_dai(struct tplg_pre_processor *tplg_pp,
+		tplg_pp_debug("set_dai_nhlt found DMIC");
+		ret = nhlt_dmic_set_params(tplg_pp, cfg, parent);
+		if (ret < 0)
+			return ret;
```
ok

juimonen commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -215,6 +217,15 @@ int nhlt_create(struct tplg_pre_processor *tplg_pp)
 
 	/* add endpoints here, increase eps_count */
 
+	/* we always have only 0 or 1 dmic ep */
```
ok

juimonen commented on topology/nhlt/intel/dmic/dmic-debug.c:
```diff
@@ -0,0 +1,180 @@
+	for (j = 0; j < DMIC_HW_CONTROLLERS; j++) {
+		tplg_pp_debug("fir_coeffs a length %u:", dmic->dmic_fir_array.fir_len[0]);
+		lines = dmic->dmic_fir_array.fir_len[0] / 8;
+		remind = dmic->dmic_fir_array.fir_len[0] % 8;
```
heh yep, will change

juimonen commented on topology/nhlt/intel/dmic/dmic-debug.c:
```diff
@@ -0,0 +1,180 @@
+	char *nhlt_p = &nhlt_string_buffer[0];
+	int i;
+
+	for (i = 0; i < size; i++) {
```
ok

juimonen commented on topology/nhlt/intel/dmic/dmic-debug.c:
```diff
@@ -0,0 +1,180 @@
+	char *nhlt_p = &nhlt_string_buffer[0];
+	int i;
+
+	for (i = 0; i < size; i++) {
```
ok

juimonen commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+/* This is a divide function that returns ceil of the quotient. E.g. ceil_divide(9, 3) returns 3,
+ * ceil_divide(10, 3) returns 4.
+ */
+static int ceil_divide(int a, int b)
```
hmm I think that is for floating point values, this is for fixed point. This is lifted from sof, and by Seppo, so don't want to too much touch it...

juimonen commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+	return c;
+}
+
+/* This function searches from vec[] (of length vec_length) integer values of n. The indexes to
```
ok

juimonen commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+
+/* This function searches from vec[] (of length vec_length) integer values of n. The indexes to
+ * equal values is returned in idx[]. The function returns the number of found matches.
+ * The max_results should be set to 0 (or negative) or vec_length get all the matches. The
```
ok

juimonen commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+			     dmic->dmic_prm[di].io_clk / fs / 2 -
+			     DMIC_FIR_PIPELINE_OVERHEAD);
+
+	i = 0;
```
ok

juimonen commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+	new_amax = Q_MULTSR_32X32((int64_t)amax, fir_gain, 31,
+				  DMIC_FIR_SCALE_Q, DMIC_FIR_SCALE_Q);
+	if (new_amax <= 0)
+		return -1;
```
ok, as this was lifted from sof -> didn't want to change too much -> but will change

juimonen commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1379 @@
+	 */
+	if (modes->num_of_modes == 0) {
+		fprintf(stderr, "select_mode(): no modes available");
+		return -1;
```
ok

juimonen commented:
@ranj063 easy comments fixed, I need to little bit think&test the open ones...

lgirdwood commented:
@libinyang fyi - PR for the NHLT work you are using.

lgirdwood commented:
@juimonen is this still WIP ?
@libinyang have you been able to fully validate ?

juimonen commented:
@lgirdwood we should pretty close, I'll do final checking and remove the WIP end of week...

libinyang commented:
> @juimonen is this still WIP ? @libinyang have you been able to fully validate ?

My last status is: I tested it on tgl-h platform. DMIC works while SSP not. And I don't have change to hear the real recorded file. I will check the sound today and update the status.

libinyang commented:
> > @juimonen is this still WIP ? @libinyang have you been able to fully validate ?
> 
> My last status is: I tested it on tgl-h platform. DMIC works while SSP not. And I don't have change to hear the real recorded file. I will check the sound today and update the status.

It seems my tgl-h platform is something wrong. DMIC blob only records the noise. And I test the daily built images on my tgl-h platform, there is still only noise. So it should not be the DMIC blob issue.

juimonen commented:
@RanderWang can you review the SSP calculation in this PR? It is a "mix" from sof fw code and your ssp blob generation code (so it has still all other SSP formats supported besides TDM/DSP_B).

plbossart commented:
@juimonen is there a specific reason why you're using the 'sof' master instead of the upstream one? 4 out of 6 commits are upstream already.

plbossart commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,321 @@
+
+	/* we don't have endpoints */
+	if (!eps_count)
+		return 0;
```
eps_count is not modified since line 240. use cppcheck or something on this code :-)

plbossart commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,321 @@
+
+err:
+	/* remove all enpoints */
+	for (i = 0; i < eps_count; i++)
```
eps_count is still zero, how does this work?

plbossart commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,321 @@
+
+	/* get blob total size */
+	nhlt_size = sizeof(struct nhlt);
+	for (i = 0; i < eps_count; i++) {
```
eps_count is zero!

plbossart commented on topology/nhlt/nhlt.h:
```diff
@@ -0,0 +1,206 @@
+struct specific_config
+{
+	uint32_t capabilities_size; /* does not include size of this field */
+	uint8_t capabilities[0];
```
I thought [] is preferred?

juimonen replied:
yes, will change

plbossart commented on topology/nhlt/nhlt.h:
```diff
@@ -0,0 +1,206 @@
+	struct device_specific_config device_config;
+	uint8_t array_type_ex;
+	uint8_t number_of_microphones;
+	uint8_t mic_vendor_configs[0];
```
[]

juimonen replied:
ok

plbossart commented on topology/nhlt/nhlt.h:
```diff
@@ -0,0 +1,206 @@
+struct formats_config
+{
+	uint8_t formats_count;
+	uint8_t f_configs[0];
```
[]

juimonen replied:
ok

plbossart commented on topology/nhlt/nhlt.h:
```diff
@@ -0,0 +1,206 @@
+{
+	struct efi_acpi_description_header efi_acpi;
+	uint8_t endpoint_count;
+	uint8_t endpoints[0];
```
[]

juimonen replied:
ok

plbossart commented on topology/nhlt/intel/dmic/dmic-debug.c:
```diff
@@ -0,0 +1,197 @@
+void dmic_print_bytes_as_hex(uint8_t *src, size_t size)
+{
+	tplg_pp_debug("printing dmic vendor blob as bytes:");
+	char nhlt_string_buffer[MAX_NHLT_BUFFER_SIZE];
```
25K on the stack?

juimonen replied:
removed

plbossart commented on topology/nhlt/intel/ssp-nhlt.c:
```diff
@@ -0,0 +1,600 @@
+  All rights reserved.
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of version 2 of the GNU General Public License as
```
"Ssp code is lifted from Sound Open Firmware (sof) code base, thus it
will have BSD-3 license."

???

juimonen replied:
"ssp _blob generation_ code is lifted" then maybe... so the intel specific blob thing is from sof, generic acpi nhlt is just new from me.

plbossart commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -265,6 +267,14 @@ int nhlt_create(struct tplg_pre_processor *tplg_pp)
+			goto err;
+		eps_count++;
+	}
+
```
you should really have a comment in the earlier code that tells eps_count is modified in future patches... Or move the eps_count to when it's actually used, otherwise static analysis will have a field day.

juimonen replied:
hmm yes, I'll run the patches through some tool and try to fix this

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,321 @@
+	int ret;
+
+	/* add the nhlt bytes as new config into existing SectionData*/
+	ret = snd_config_search(tplg_pp->output_cfg, "SectionData", &top_data_section);
```
can we use tplg_create_config_template() to create the Data config with bytes in it?

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,321 @@
+			continue;
+
+		/* item already exists */
+		if (!strcmp(name, data_name))
```
how does this work? so you have one NHLT blob for types of DAI?

juimonen replied:
it is just 1 acpi nhlt blob, that might contain 1 to many nhlt endpoints (DAI configs). So I'm just trying to add 1 data element into manifest, nothing more. I think the code is a copy from some other part from topology2 processing, can be changed if there's standard way of adding data element.

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -0,0 +1,321 @@
+		return -EINVAL;
+
+	if(tplg_pp->private_data)
+		free(tplg_pp->private_data);
```
you dont seem to allocate preivate_data in init()?

juimonen replied:
Allocated in nhlt_dmic_init_params(tplg_pp) or nhlt_ssp_init_params(tplg_pp). Not very clear, but I' trying to separate the nhlt and the blob generation (headers and code) from each other.

ranj063 commented on topology/nhlt/intel/dmic/dmic-debug.c:
```diff
@@ -0,0 +1,197 @@
+	}
+
+	for (i = 0; i < size; i++) {
+		snprintf(nhlt_p, 6, "0x%02x,", *src);
```
do we really need the snprintf , why cant we directly do a printf?

juimonen replied:
I mean this is to print the internal byte values as alsa conf hex bytes into a string (which is then added to the conf)?

juimonen replied:
of yes here in the debug, yes sorry will change to printf. This was just mimicking exactly same code as in actual nhlt blob generation in upper levels, will change.

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -253,6 +256,15 @@ int nhlt_create(struct tplg_pre_processor *tplg_pp)
+		/* the index is always 0 in dmic case */
+		ret = nhlt_dmic_get_ep(tplg_pp, &eps[eps_count], 0);
+		if (ret < 0)
+			return ret;
```
you need to unwind and free previously successful ep alloc's right?

juimonen replied:
there isn't any ep's allocated before this...

ranj063 commented on topology/nhlt/nhlt-processor.c:
```diff
@@ -316,6 +329,11 @@ int nhlt_set_dai(struct tplg_pre_processor *tplg_pp,
 
 	/* set dai parameters here */
+	if (!strncmp(id, "DMIC", 4)) {
+		ret = nhlt_dmic_set_params(tplg_pp, cfg, parent);
```
what does cfg point to here?

ranj063 replied:
does nhlt_dmic_set_params() end up with generating the blob? Its hard to tell where the blob generation is happening

juimonen replied:
cfg should be the DAI config from topology?

juimonen replied:
I'm getting the dai params from topology2 parsing (with auto_attr) -> setting the values to internal structs -> all blobs are generated at the end of parsing from these internal structs. Sof fw code is doing it like this -> I don't want to start to modify too much for e.g. the dmic blob code -> will differ from sof too much -> horrible amount of possible errors for me

ranj063 commented on topology/nhlt/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1002 @@
+		}
+	}
+
+	/* set object values */
```
why do you need to set class values? The object should have inherited class values already no?

juimonen replied:
so are the class values already set at this point in auto_attr processing?

ranj063 replied:
I would think so

ranj063 commented on topology/nhlt/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1002 @@
+				   polarity_b, clk_edge, skew);
+}
+
+static int set_mic_data(struct tplg_pre_processor *tplg_pp, snd_config_t *cfg)
```
you call this a set but its really a get isnt it? you're getting the user values from the input conf file and updating your local stuct for processing later

juimonen replied:
heh well, I guess it could be thought both ways. I was thinking it this way: "I'm setting the data to internal structs from topology" then "getting the calculated endpoints from the engine". So I see the nhlt part as an external engine where you need to set data, then ask it to generate the blobs

ranj063 commented on topology/nhlt/intel/dmic-nhlt.c:
```diff
@@ -0,0 +1,1002 @@
+ *		}
+ *      }
+ */
+int nhlt_dmic_set_params(struct tplg_pre_processor *tplg_pp,
```
same here, its really get_params and to be more precise, we should call it nhlt_dmic_get_input_params()

ranj063 commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1365 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
I wonder if it would make it easier if you split the dmic-process.* files into a separate commit. There are pretty much exactly copy from the SOF code right?

juimonen replied:
I mean whatever makes things easier to read. I've already done 2 iterations from comments to back and forth with single commits to multiple commits :) but yeah as it is not so clear -> there's probably issue in the patch division

ranj063 commented on topology/nhlt/intel/dmic/dmic-process.c:
```diff
@@ -0,0 +1,1365 @@
+	return 0;
+}
+
+static struct intel_dmic_params * get_dmic(struct tplg_pre_processor *tplg_pp)
```
I think your get and set are inverted. get should be used for reading input values form conf and set should be used to set the values in the generated blob.

juimonen commented:
> @juimonen is there a specific reason why you're using the 'sof' master instead of the upstream one? 4 out of 6 commits are upstream already.

should be now rebased on upstream.

juimonen commented:
@plbossart I ran now checkpatch, sparse and cppcheck on all patches individually. There we're a lot of style issues and couple of coding things... which should be now fixed. But could not get them to complain about the eps_count thing you mentioned :)
If you have some guide for me to run some other tools or run them in some other way?

plbossart commented:
> @plbossart I ran now checkpatch, sparse and cppcheck on all patches individually. There we're a lot of style issues and couple of coding things... which should be now fixed. But could not get them to complain about the eps_count thing you mentioned :) If you have some guide for me to run some other tools or run them in some other way?

well that's a good sign if humans are still needed. Tools are still somewhat unreliable, I've seen static analysis results show up much later due to completely unrelated changes. I don't have any specific recommendations other than use the following options

sparse:
export ARCH=x86_64 CF="-Wsparse-error -Wsparse-all -Wno-bitwise-pointer -Wno-pointer-arith -Wno-typesign -Wnoshadow"

cppcheck:
cppcheck --platform=unix64 --force --max-configs=1024 --inconclusive --enable=all --suppress=variableScope --suppress=shiftTooManyBitsSigned --suppress=arithOperationsOnVoidPointer --suppress=bitwiseOnBoolean *.c
